syntax = "proto3";

package pagi.v1;

option go_package = "github.com/pagi-ai/pagi/contracts/gen/go/pagiv1;pagiv1";
option java_package = "com.pagi.contracts.v1";
option java_multiple_files = true;

// Canonical, provider-agnostic request surface.
//
// This schema is intentionally richer than a minimal "prompt". It supports:
// - message history normalization
// - multimodal content
// - tool calling and strict schemas
// - generation constraints
// - metadata for routing/observability

enum MessageRole {
  MESSAGE_ROLE_UNSPECIFIED = 0;
  MESSAGE_ROLE_SYSTEM = 1;
  MESSAGE_ROLE_USER = 2;
  MESSAGE_ROLE_ASSISTANT = 3;
  MESSAGE_ROLE_TOOL = 4;
}

message TextPart {
  string text = 1;
}

message ImagePart {
  string url = 1;
}

message AudioPart {
  string url = 1;
}

message FilePart {
  string url = 1;
  string mime_type = 2; // e.g. application/pdf
}

message ContentPart {
  oneof part {
    TextPart text = 1;
    ImagePart image = 2;
    AudioPart audio = 3;
    FilePart file = 4;
  }
}

message Message {
  MessageRole role = 1;
  repeated ContentPart content = 2;
  string name = 3;
  string tool_call_id = 4;
}

message Tool {
  string name = 1;
  string description = 2;
  string parameters_json_schema = 3; // JSON Schema as a string
  bool strict = 4;
}

message GenerationConstraints {
  uint32 max_tokens = 1;
  float temperature = 2;
  float top_p = 3;
  uint32 top_k = 4;
  repeated string stop_sequences = 5;
  float presence_penalty = 6;
  float frequency_penalty = 7;
  string reasoning_effort = 8;
  bool stream = 9;
}

message CanonicalAIRequest {
  string request_id = 1;
  string agent_id = 2;
  string session_id = 3;
  repeated Message messages = 4;
  repeated Tool tools = 5;
  string tool_choice = 6;
  GenerationConstraints constraints = 7;
  string preferred_model = 8;
  map<string, string> metadata = 9;
  string response_format_json_schema = 10;
}

message CanonicalAIResponse {
  string request_id = 1;
  string adapter_id = 2;
  string json = 3;
}

message AdapterCapabilities {
  bool streaming = 1;
  bool token_count = 2;
  bool model_route = 3;
  bool embed_cache = 4;
}

message RegisterAdapterRequest {
  string adapter_id = 1;
  string endpoint = 2; // e.g. http://127.0.0.1:6000
  AdapterCapabilities capabilities = 3;
  string version = 4;
}

message RegisterAdapterResponse {
  bool ok = 1;
}

message ListAdaptersRequest {}

message AdapterInfo {
  string adapter_id = 1;
  string endpoint = 2;
  AdapterCapabilities capabilities = 3;
  string version = 4;
}

message ListAdaptersResponse {
  repeated AdapterInfo adapters = 1;
}

service AdapterRegistry {
  rpc Register(RegisterAdapterRequest) returns (RegisterAdapterResponse);
  rpc List(ListAdaptersRequest) returns (ListAdaptersResponse);
}

service AdapterService {
  rpc Process(CanonicalAIRequest) returns (CanonicalAIResponse);
}

